federation/pkg/federation-controller/service/dns/dns.go

```diff
@@ -0,0 +1,431 @@
+	rrsets, supported := dnsZone.ResourceRecordSets()
+	if !supported {
+		return fmt.Errorf("DNS provider does not support the ResourceRecordSets interface")
+	}
+
+	changeSet := rrsets.StartChangeset()
+
+	for dnsName, endpoints := range endpoints {
+		rrset, err := rrsets.Get(dnsName)
+		if err != nil {
+			return err
+		}
+
+		if len(endpoints) < 1 {
+			// There are no healthy endpoints for the dnsName, So need to add an appropriate CNAME record.
+			uplevelCname := d.getUplevelCName(dnsName, dnsNamePrefix)
+			glog.V(4).Infof("There are no healthy endpoint addresses at level %q, so CNAME to %q, if provided", dnsName, uplevelCname)
+			if uplevelCname != "" {
+				// dnsName is for zone or region level
+				// should ensure a CNAME pointing to one level above
+				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
+				if rrset == nil {
+					// rrset does not exist, so need to create new CNAME Record
+					glog.V(4).Infof("Creating CNAME to %q for %q", uplevelCname, dnsName)
+					changeSet.Add(newRrset) // the rrset already exists, ensure that it is right.
+
+				} else {
+					if !dnsprovider.ResourceRecordSetsEquivalent(rrset, newRrset) {
+						glog.V(4).Infof("Have recordset %v. Need recordset %v", rrset, newRrset)
+						changeSet.Remove(rrset).Add(newRrset)
+					} else {
+						glog.V(4).Infof("Existing recordset %v is equivalent to needed recordset %v, our work is done here.", rrset, newRrset)
+					}
+				}
+			} else {
+				// dnsName is for global level
+				if rrset == nil {
+					glog.V(4).Infof("We want no record for %q, and we have no record, so we're all good.", dnsName)
+				} else {
+					// the rrset exists, so should remove it as no healthy endpoints exist now.
+					glog.V(4).Infof("Successfully removed existing recordset %v", rrset)
+					changeSet.Remove(rrset)
+				}
+			}
+		} else {
+			// We have valid endpoint addresses, so just add them as A records.
+			// But first resolve DNS names, as some cloud providers (like AWS) expose
+			// load balancers behind DNS names, not IP addresses.
+			glog.V(4).Infof("We have valid endpoint addresses %v at level %q, so add them as A records, after resolving DNS names", endpoints, dnsName)
+			resolvedEndpoints, err := getResolvedEndpoints(endpoints)
+			if err != nil {
+				return err // TODO: We could potentially add the ones we did get back, even if some of them failed to resolve.
+			}
+			newRrset := rrsets.New(dnsName, resolvedEndpoints, minDnsTtl, rrstype.A)
+			if rrset == nil {
+				// rrset does not exist, so need to create new A Record
+				glog.V(4).Infof("Adding recordset %v for %s", newRrset, dnsName)
+				changeSet.Add(newRrset)
+			} else {
+				// the rrset already exists, ensure that it is right.
+				if !dnsprovider.ResourceRecordSetsEquivalent(rrset, newRrset) {
+					// TODO: We could be more thorough about checking for equivalence to avoid unnecessary updates, but in the
```

Comment:
I don't think that this comment applies any longer.  We are now being thorough about checking for equivalence?